client/src/main/java/io/pravega/client/stream/impl/ClientFactoryImpl.java

```diff
@@ -178,15 +179,38 @@ public ClientFactoryImpl(String scope, Controller controller, ConnectionFactory
         Consumer<Segment> segmentSealedCallBack = s -> {
             throw new IllegalStateException("RevisionedClient: Segmentsealed exception observed for segment:" + s);
         };
-        String delegationToken = Futures.getAndHandleExceptions(controller.getOrRefreshDelegationTokenFor(segment.getScope(),
-                segment.getStreamName()), RuntimeException::new);
+
+        String delegationToken = checkStreamReturnDelegation(streamName, segment);
+
         SegmentOutputStream out = outFactory.createOutputStreamForSegment(segment, segmentSealedCallBack,
                 config.getEventWriterConfig(), delegationToken);
         ConditionalOutputStream cond = condFactory.createConditionalOutputStream(segment, delegationToken, config.getEventWriterConfig());
         SegmentMetadataClient meta = metaFactory.createSegmentMetadataClient(segment, delegationToken);
         return new RevisionedStreamClientImpl<>(segment, in, out, cond, meta, serializer);
     }
 
+    /**
+     * Check the stream state and get the delegation string; if state of stream is not UNKNOWN or SEALED, continue
+     * to retrieve delegation for given stream, otherwise throw an exception.
+     * @param streamName Stream name
+     * @param segment    Segment
+     * @return A delegation string used for RevisionedStreamClientImpl creation or throw exception if the Stream is not ready.
+     */
+    private String checkStreamReturnDelegation(String streamName, Segment segment) {
+        CompletableFuture<StreamState>  streamState = controller.getStreamState(scope, streamName);
+        CompletableFuture<String> delegationToken = controller.getOrRefreshDelegationTokenFor(segment.getScope(), segment.getStreamName());
+        return streamState
+                .thenCombine(delegationToken,
+                        (state, delTok) -> {
+                            if (state != StreamState.UNKNOWN && state != StreamState.SEALED) {
+                                return delTok;
+                            } else {
+                                throw new IllegalStateException("Stream is not ready: " + streamName);
```

Comment:
This is a public facing error message. Let's make it as clear as possible. For example we should probably throw `InvalidStreamException` and include the stream name and why it is invalid. (For example "Does not exist in the scope: ")